crates/wasmparser/src/validator/component.rs

```diff
@@ -1048,6 +1086,896 @@ impl ComponentState {
         Ok(())
     }
 
+    pub fn task_backpressure(
+        &mut self,
+        types: &mut TypeAlloc,
+        offset: usize,
+        features: &WasmFeatures,
+    ) -> Result<()> {
+        if !features.component_model_async() {
+            bail!(
+                offset,
+                "`task.backpressure` requires the component model async feature"
+            )
+        }
+
+        let (_is_new, group_id) = types.intern_canonical_rec_group(
+            false,
+            RecGroup::implicit(
+                offset,
+                SubType {
+                    is_final: true,
+                    supertype_idx: None,
+                    composite_type: CompositeType {
+                        shared: false,
+                        inner: CompositeInnerType::Func(FuncType::new([ValType::I32], [])),
+                    },
+                },
+            ),
+        );
+        let id = types[group_id].start;
+        self.core_funcs.push(id);
+        Ok(())
+    }
+
+    pub fn task_return(
+        &mut self,
+        type_index: u32,
+        offset: usize,
+        features: &WasmFeatures,
+    ) -> Result<()> {
+        if !features.component_model_async() {
+            bail!(
+                offset,
+                "`task.return` requires the component model async feature"
+            )
+        }
+
+        let id = self.type_id_at(type_index, offset)?;
+        self.core_funcs.push(id);
```

Comment:
Should this validate that `id` points to a core function type? (as opposed to an array/struct/etc)

Also this may be elsewhere but I would otherwise have expected some sort of validation of the type here as well (as opposed to allowing any signature)

Reply:
Regarding the first point: yes, I'll add a check to make sure it's a core function type.

Regarding the second point: We actually do need to allow arbitrary signatures here, since the signature will vary depending on which component-level exported function the module is calling `task.return` for, and we can't statically determine that in general, so we need to delay the type check until runtime (at which point we've presumably interned the expected and actual type, so it's just a quick integer comparison).